### PR TITLE
[bazel] Add clangd as a library support

### PR DESCRIPTION
**Pull request by josh11b**
Creates a `BUILD.bazel` file for the `clangd` directory in the project overlay.

This upstreams the patch that allows https://github.com/carbon-language/carbon-lang/tree/trunk/language_server to use `clangd` as a library. This was created as part of a Summer of Code project building a prototype Carbon language server. If this is not an appropriate architecture, I'm very open to alternative paths forward.

Thanks!

### CODE REVIEW

**Comment by josh11b** [comment]
I've made my email address public, as requested.

**Comment by josh11b** [comment]
Ping!

File: utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+package(default_visibility = ["//visibility:public"])
+
+licenses(["notice"])
+
+cc_library(
+    name = "clangd_library",
```
Comment (MaskRay) [inline utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel:10]: clangDaemon would match the library name in clang-tools-extra/clangd/CMakeLists.txt

**Comment by josh11b** [comment]
Ping!

File: utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+package(default_visibility = ["//visibility:public"])
```
Comment (MaskRay) [inline utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel:5]: Add `features = ["layering_check"],`

File: utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel (hunk 1 above)
Comment (josh11b) [reply utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel]: Renamed

File: utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel (hunk 2 above)
Comment (josh11b) [reply utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel]: Added

File: utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        "index/SymbolID.h",
+        "support/Function.h",
+        "support/Cancellation.h",
+        "support/ThreadCrashReporter.h",
+        "support/Logger.h",
+        "support/Trace.h",
+        "support/MemoryTree.h",
+        "support/Context.h",
+        "support/Shutdown.h",
+    ],
```
Comment (rupprecht) [inline utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel:39]: clangd/CMakeLists.txt lists many other files needed by the clangDaemon target.

I think this is totally fine to submit as-is, as maybe this is the minimal set of files that one needs for basic clangd functionality, but we should probably have a TODO to pick up other files (e.g. `glob(["*.cpp", "dir/**/*.cpp", ...])`) for more complete functionality. (I'm not a clangd expert, so I'm not sure what the implication of leaving out all those other files is).

**Comment by MaskRay** [comment]
> Ping!

llvm-project email notifications are extremely high-volume and nobody can notice every change even if they are subscribed to the relevant labels...

If you don't have write/triage role (https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/managing-repository-roles/repository-roles-for-an-organization#repository-roles-for-organizations), the best is to CC relevant people (`git log -- directory` and finding recent reviewers is a good way).

(Now, I noticed this patch because I noticed a message in an IM I am subscribed to... not that I actively check clang-tools-extra patches/issues)

Feel free to ping the patch once a week when no feedback is received per https://llvm.org/docs/Contributing.html#how-to-submit-a-patch

> If you have received no comments on your patch for a week, you can request a review by ‘ping’ing the GitHub PR with “Ping”. The common courtesy ‘ping’ rate is once a week. Please remember that you are asking for valuable time from other professional developers.

File: utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,46 @@
+# This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+package(features = ["layering_check"],
```
Comment (MaskRay) [inline utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel:5]: llvm/BUILD.bazel places `default_visibility` before `features`. Best to follow its convention.

File: clang-tools-extra/clangd/Transport.h [hunk 5]
```diff
@@ -18,7 +18,6 @@
 #ifndef LLVM_CLANG_TOOLS_EXTRA_CLANGD_TRANSPORT_H
 #define LLVM_CLANG_TOOLS_EXTRA_CLANGD_TRANSPORT_H
 
-#include "Feature.h"
```
Comment (MaskRay) [inline clang-tools-extra/clangd/Transport.h:21]: This change should be removed.

**Review by MaskRay — APPROVED** [review]
bazel-specific changes often have a title `[bazel] Add clangd support`. Please consider following suit.

It would be nice to have a more complete support for clangd, but I guess having this cc_library first can make your downstream changes easier, so LGTM.

File: utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel (hunk 1 above)
Comment (rupprecht) [reply utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel]: Small nit: IIUC the convention in the llvm bzl tree is to use casing like `ClangDaemon`.

File: clang-tools-extra/clangd/Transport.h (hunk 5 above)
Comment (josh11b) [reply clang-tools-extra/clangd/Transport.h]: Done.

File: utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel (hunk 4 above)
Comment (josh11b) [reply utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel]: Done.

File: utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel (hunk 3 above)
Comment (josh11b) [reply utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel:42]: I added a # TODO comment.

File: utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel (hunk 1 above)
Comment (josh11b) [reply utils/bazel/llvm-project-overlay/clang-tools-extra/clangd/BUILD.bazel]: Done.

**Review by josh11b — COMMENTED** [review]
Thank you for the feedback!

**Review by rupprecht — APPROVED** [review]

**Comment by rupprecht** [comment]
Thanks! Sorry about the delay, we should get the bzl tree setup for automatic assignment to reviewers.

**Pull request merged**